builtin/providers/aws/resource_aws_volume_attachment.go

```diff
@@ -0,0 +1,158 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/aws/awserr"
+	"github.com/awslabs/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsVolumeAttachment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsVolumeAttachmentCreate,
+		Read:   resourceAwsVolumeAttachmentRead,
+		Delete: resourceAwsVolumeAttachmentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"device_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"instance_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"volume_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"force_detach": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsVolumeAttachmentCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	name := d.Get("device_name").(string)
+	iID := d.Get("instance_id").(string)
+	vID := d.Get("volume_id").(string)
+
+	opts := &ec2.AttachVolumeInput{
+		Device:     aws.String(name),
+		InstanceID: aws.String(iID),
+		VolumeID:   aws.String(vID),
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"pending"},
+		Target:     "attaching",
+		Refresh:    attachVolumeFunc(conn, opts),
+		Timeout:    1 * time.Minute,
+		Delay:      5 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err := stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error attaching volume %s to instance %s: %s", vID, iID, err)
+	}
+
+	d.SetId(volumeAttachmentID(name, vID, iID))
+	return resourceAwsVolumeAttachmentRead(d, meta)
+}
+
+func attachVolumeFunc(conn *ec2.EC2, opts *ec2.AttachVolumeInput) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		va, err := conn.AttachVolume(opts)
```

Comment:
@phinze

Currently aws_ebs_volume doesn’t wait if the volume is ready.

That makes sense that resource aws_volume_attachment doesn’t wait for the availability of the volume but that aws_ebs_volume guarantees it by its return.

-- 
Julien Lavergne
Sent with Airmail

On May 27, 2015 at 4:39:32 PM, Paul Hinze (notifications@github.com) wrote:

In builtin/providers/aws/resource_aws_volume_attachment.go:

> +              Delay:      5 * time.Second,
> +          MinTimeout: 3 * time.Second,
> +  }
> +
> +  _, err := stateConf.WaitForState()
> +  if err != nil {
> +          return fmt.Errorf("Error attaching volume %s to instance %s: %s", vID, iID, err)
> +  }
> +
> +  d.SetId(volumeAttachmentID(name, vID, iID))
> +  return resourceAwsVolumeAttachmentRead(d, meta)
> +}
> +
> +func attachVolumeFunc(conn *ec2.EC2, opts *ec2.AttachVolumeInput) resource.StateRefreshFunc {
> +  return func() (interface{}, string, error) {
> +          va, err := conn.AttachVolume(opts)
In TF's case, say that volume is another resource. It gets created, but it's create func doesn't wait for it to be "ready", just "exist".

Seems like it should be the volume's responsibility to wait to return until the volume is ready, such that subsequent resources don't have to worry about it.

So then we can expect a single call to AttachVolumes to do its thing (or be able to error right away), then we just poll to wait for its thing to be done.

Let me know if that makes sense.

—
Reply to this email directly or view it on GitHub.